rpcserver.go

```diff
@@ -7676,13 +7699,44 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
 		time.Duration(expirySeconds) * time.Second,
 	)
 
+	maxUnits := req.AssetAmount
+	amtMsat := lnwire.MilliSatoshi(0)
+
+	// If the invoice defines the desired amount in satoshis, we need to
+	// query our oracle first to get an estimation on the asset rate. This
+	// will help us establish a quote with the correct amount of asset
+	// units.
+	if satsMode {
+		amtMsat = lnwire.MilliSatoshi(req.InvoiceRequest.ValueMsat +
+			req.InvoiceRequest.Value*(1000))
+
+		oracleRes, err := r.cfg.PriceOracle.QueryBidPrice(
+			ctx, specifier, fn.None[uint64](), fn.Some(amtMsat),
+			fn.None[rfqmsg.AssetRate](),
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		if oracleRes.Err != nil {
+			return nil, fmt.Errorf("cannot query oracle: %v",
+				oracleRes.Err.Error())
+		}
```

Comment:
Would be good to start to split up some of this logic into helper functions. Looks like we have the distinct phases of: 
  * initial parsing/validation 
  * creation of the receive/buy order 
  * invoice creation